MDEV-33660 Add note when setting AUTO_INCREMENT lower than next value

FarihaIS opened the pull request:
## Description

Currently, `ALTER TABLE ... AUTO_INCREMENT = N` silently fails when using values lower than the next AUTO_INCREMENT value.

Add a note when attempting to set AUTO_INCREMENT to a lower value than the next AUTO_INCREMENT value in the column, and maintain the higher value.

Update existing test suites and add a new test to the maria suite for the newly added note.

## Release Notes

N/A

## How can this PR be tested?

Execute the main, innodb, maria, and vcol test suites in mysql-test-run. 

This commit also adds a test in the `alter` test of the maria suite (`maria.alter`).

### Before the fix

Running `maria.alter` (with the added test) fails as below:
```
CURRENT_TEST: maria.alter
--- /quick-rebuilds/mariadb/mysql-test/suite/maria/alter.result 2026-01-16 19:41:40.220876885 +0000
+++ /quick-rebuilds/mariadb/mysql-test/suite/maria/alter.reject 2026-01-16 19:57:10.394116392 +0000
@@ -191,11 +191,8 @@
 INSERT INTO t1 VALUES (1),(2),(3),(4),(5);
 # Attempt to set AUTO_INCREMENT to a lower value
 ALTER TABLE t1 AUTO_INCREMENT=3;
-Warnings:
-Note	4190	ALTER TABLE is implicitly changing the value of 'AUTO_INCREMENT' from '3' to '6'
 SHOW WARNINGS;
 Level  Code    Message
-Note	4190	ALTER TABLE is implicitly changing the value of 'AUTO_INCREMENT' from '3' to '6'
 SHOW CREATE TABLE t1;
 Table  Create Table
 t1     CREATE TABLE `t1` (
@@ -222,11 +219,8 @@
 4
 # Attempt to set AUTO_INCREMENT to a lower value
 ALTER TABLE t1 AUTO_INCREMENT=3;
-Warnings:
-Note	4190	ALTER TABLE is implicitly changing the value of 'AUTO_INCREMENT' from '3' to '5'
 SHOW WARNINGS;
 Level  Code    Message
-Note	4190	ALTER TABLE is implicitly changing the value of 'AUTO_INCREMENT' from '3' to '5'
 SHOW CREATE TABLE t1;
 Table  Create Table
 t1     CREATE TABLE `t1` (

Result length mismatch
```

### After the fix

Running `maria.alter` (with the added test) succeeds:
```
maria.alter                              [ pass ]     97
```

## Basing the PR against the correct MariaDB version

* [x] _This is a bug fix, and the PR is based against the branch 10.11._

## Copyright

All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

FarihaIS committed:
MDEV-33660 Add note when setting AUTO_INCREMENT lower than next value

Currently, ALTER TABLE ... AUTO_INCREMENT = N silently fails when using
values lower than the next AUTO_INCREMENT value in the column.

Add a note when attempting to set AUTO_INCREMENT to a lower value than
the next AUTO_INCREMENT value, and maintain the higher value.

Update existing test suites and add a new test to the maria suite for
the newly added note.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MariaDB/server?pullRequest=4254) <br/>All committers have signed the CLA.

FooBarrior commented on storage/maria/ha_maria.cc:
```diff
@@ -3220,6 +3222,21 @@ void ha_maria::update_create_info(HA_CREATE_INFO *create_info)
+          ER_ALTER_INFO,
+          "Can not set AUTO_INCREMENT to %" PRIu64 " which is lower than the current max value of %" PRIu64,
+          static_cast<uint64_t>(create_info->auto_increment_value),
+          static_cast<uint64_t>(current_max));
```
No need for a `static_cast`

FooBarrior commented on storage/maria/ha_maria.cc:
```diff
@@ -3220,6 +3222,21 @@ void ha_maria::update_create_info(HA_CREATE_INFO *create_info)
   {
     create_info->auto_increment_value= stats.auto_increment_value;
   }
+  else if (thd_sql_command(ha_thd()) == SQLCOM_ALTER_TABLE)
```
handler::update_create_info is also called from show_create_table, so you'll end up in a warning for every SHOW CREATE TABLE

FooBarrior commented on mysql-test/suite/maria/alter.result:
```diff
@@ -182,6 +182,35 @@ pk	a
 #
+# MDEV-33660 Add warning when setting AUTO_INCREMENT lower than max value
+#
+CREATE TABLE t1 (id INT AUTO_INCREMENT PRIMARY KEY) ENGINE=Aria;
```
Add CREATE TABLE ... AUTO_INCREMENT=3 test

FooBarrior reviewed (changes requested):
> How can this PR be tested?
I think the correct answer is running the test you have provided within commit. Would be embarrasing to make every our contributor fill a whole server setup instruction in a PR description

> Could you please guide me to where I might add the change to address all engines if this is possible?
I'd suggest checking somewhere both ALTER and CREATE call. This place can be one of the functions called from mysql_prepare_create_table.

FarihaIS commented on storage/maria/ha_maria.cc:
```diff
@@ -3220,6 +3222,21 @@ void ha_maria::update_create_info(HA_CREATE_INFO *create_info)
+  else if (thd_sql_command(ha_thd()) == SQLCOM_ALTER_TABLE)
+  {
+    // If this is ALTER TABLE and a new AUTO_INCREMENT value was specified
+    uint64_t current_max= stats.auto_increment_value;
```
I had replaced the `ulonglong` with the fixed-size type in standard C++ lib as per this Jira https://jira.mariadb.org/browse/MDEV-35460 since it will make the refactoring slightly easier later on. 

If you feel we can postpone this replacement and would like me to change it back to `ulonglong`, please let me know.

vuvova reviewed (changes requested):
1. don't do it in 10.6, please. rebase your PR 11.4 please.

2. The approach looks wrong. Try this test case:
```sql
create table t1 (b int);
insert t1 values (1),(2),(3);
alter table t1 add column a int primary key auto_increment, auto_increment=100;
select * from t1;
```

gkodinov reviewed (approved):
Thank you for your contribution. This is a preliminary review.

LGTM. 

Please stand by for the final review.

vuvova commented on sql/sql_table.cc:
```diff
@@ -131,6 +131,44 @@ static bool
+{
+  if (!(create_info->used_fields & HA_CREATE_USED_AUTO) ||
+      create_info->auto_increment_value == 0 ||
+      !table || !table->found_next_number_field || !table->file)
```
this is **very** conservative. What does it mean?
* `!(create_info->used_fields & HA_CREATE_USED_AUTO)` — a user did not use `AUTO_INCREMENT=N` clause, sure it can happen
* `create_info->auto_increment_value == 0` — what does it mean, `AUTO_INCREMENT=0` ?
* `!table` — is it even possible here?
* `!table->found_next_number_field` — the table has no auto-increment column. Is it possible here or the earlier checks must've prevented that?
* `!table->file` — is it even possible here?

FarihaIS replied:
Good point, I removed `!table` and `!table->file` - at both call sites, the table is open with a valid handler.

I did keep the rest, however:
- `!(create_info->used_fields & HA_CREATE_USED_AUTO)` - user didn't specify `AUTO_INCREMENT=N`, which is possible
- `create_info->auto_increment_value == 0` - reachable when the `auto_inc` column is dropped (`mysql_prepare_alter_table` sets value to 0 with `HA_CREATE_USED_AUTO` in that path)
- `!table->found_next_number_field` — covers `ALTER TABLE t1 DROP COLUMN id, AUTO_INCREMENT=5` where the user-specified value is non-zero but the resulting table has no `auto_inc` column

Do the remaining checks make sense now, or are they still too conservative?

vuvova commented on sql/sql_table.cc:
```diff
@@ -131,6 +131,44 @@ static bool
+    return;
+
+  if (table->file->info(HA_STATUS_AUTO))
+    return;
```
what would that failure mean?

FarihaIS replied:
On second thought, not much I suppose. 

Removed the early return on `info(HA_STATUS_AUTO)` failure as the existing call in `mysql_prepare_alter_table` doesn't check the return value either.

vuvova commented on sql/sql_table.cc:
```diff
@@ -8222,6 +8260,8 @@ static bool mysql_inplace_alter_table(THD *thd,
     commit_succeded_with_error= 1;
   }
 
+  check_auto_increment_lower_than_max(thd, ha_alter_info->create_info, table);
```
why did you put this check in three different places, could it be possible to have it just once on a common code path?

e.g. see where `HA_STATUS_AUTO` is also used

FarihaIS replied:
Thanks for the catch, I reduced the number of call sites from 3 to 2 by removing the no-op inplace call — `AUTO_INCREMENT=N` sets `ALTER_OPTIONS`, so it never enters the no-op path. 

The remaining two (inplace and copy) operate on different table objects with potentially different `auto_increment values`, so I'm not sure how to merge them, if that's possible. 

Could you please let me know if this is okay, or if there is a better way to integrate them?

vuvova reviewed (changes requested):
targeting 11.4 looks strange. a bug fix normally goes into the earliest applicable version, a feature goes into main. So one can expect the target version to be 10.11 or 13.1, why it is 11.4?

and see comments below

FarihaIS commented:
@vuvova Thank you for the feedback. I have addressed/responded to all your comments. Could you please take a look and let me know if the PR looks fine now?

For reference, I had rebased this on branch 11.4 as per your previous suggestion here (before that, it was rebased on 10.6): https://github.com/MariaDB/server/pull/4254#pullrequestreview-3711285626

As per your latest comment however, I have just rebased it on 10.11.

FarihaIS commented:
@vuvova I noticed you left some new comments on [MDEV-33660](https://jira.mariadb.org/browse/MDEV-33660) along with some others.

It sounds like the consensus is to make this an error (downgraded to warning with `IGNORE_BAD_TABLE_OPTIONS`) targeting 13.0 only. Before I rework this PR, could you please confirm:
1. Should I rebase this onto the `main` branch then?
2. Should I use `ER_BAD_OPTION_VALUE` or a new error code?
3. Should the `ALTER` fail (return error) when the value can't be applied, or still succeed silently but report an error/warning one can see via `SHOW WARNINGS`?

vuvova commented:
Thanks for asking! Indeed, there seems to be a consensus, but not yet a decision "we do it this way" so any coding would be a bit premature now. I'll confirm in the [MDEV-33660](https://jira.mariadb.org/browse/MDEV-33660) when we'll have it.

gkodinov commented:
> Thanks for asking! Indeed, there seems to be a consensus, but not yet a decision "we do it this way" so any coding would be a bit premature now. I'll confirm in the [MDEV-33660](https://jira.mariadb.org/browse/MDEV-33660) when we'll have it.

@FarihaIS I believe we have the needed action plan now in the Jira:

> Okay. Let's issue a warning as the original description said. Or, more precisely, a note, because a warning is actionable, while a note is informational. Also warnings becomes an error in the strict mode and we don't want that, do we?


Can you please amend your change?

FarihaIS commented:
@gkodinov thank you for clarifying, I have made the changes requested in the Jira and also updated the PR title and description to match this - could you please let me know if everything looks okay?

gkodinov commented:
> @gkodinov thank you for clarifying, I have made the changes requested in the Jira and also updated the PR title and description to match this - could you please let me know if everything looks okay?

I'm just the messenger here. It's for @vuvova  to review now.

vuvova commented on mysql-test/suite/maria/alter.result:
```diff
@@ -185,6 +185,63 @@ SET GLOBAL aria_max_sort_file_size= @max_size.save;
 #
 #
+# MDEV-33660 Add warning when setting AUTO_INCREMENT lower than max value
+#
```
Actually, the header should match MDEV summary exactly. So it must be
```
---echo # MDEV-33660 ALTER TABLE ... AUTO_INCREMENT = NN; should return a Warning if it can not use the provided value
```

FarihaIS replied:
Done, updated to match the MDEV summary exactly.

vuvova commented on mysql-test/suite/maria/alter.test:
```diff
@@ -193,6 +193,39 @@ SET GLOBAL aria_max_sort_file_size= @max_size.save;
 --echo # End of 10.4 test
 --echo #
 
+--echo #
+--echo # MDEV-33660 Add warning when setting AUTO_INCREMENT lower than max value
+--echo #
```
same as above. Must match MDEV summary in Jira

FarihaIS replied:
Done, updated to match the MDEV summary exactly.

vuvova commented on sql/sql_table.cc:
```diff
@@ -127,6 +127,42 @@ static bool
+
+  MDEV-33660: When ALTER TABLE ... AUTO_INCREMENT=N is used with N lower
+  than the next AUTO_INCREMENT value, the engine keeps the higher value.
+  This function checks and issues a note if so.
```
No need to put MDEV numbers and summaries in the code. Please drop this paragraph

FarihaIS replied:
Thank you, dropped the `MDEV` reference from the comment.

vuvova commented on sql/sql_table.cc:
```diff
@@ -127,6 +127,42 @@ static bool
+    push_warning_printf(thd, Sql_condition::WARN_LEVEL_NOTE,
+                        ER_ALTER_INFO,
+                        "Can not set AUTO_INCREMENT to %llu which is lower "
+                        "than the next AUTO_INCREMENT value of %llu",
```
instead of hard-coding a message text in English, I suggest to use `WARN_OPTION_CHANGING`.
It could look like
```c++
push_warning_printf(thd, Sql_condition::WARN_LEVEL_NOTE,
    WARN_OPTION_CHANGING, "ALTER TABLE", "AUTO_INCREMENT",
    llstr(create_info->auto_increment_value, buf1), llstr(next_auto_inc, buf2));
```
yielding something like
```
ALTER TABLE is implicitly changing the value of 'AUTO_INCREMENT' from '1' to '6'
```

FarihaIS replied:
Thank you, switched to `WARN_OPTION_CHANGING` so the note now reads: `ALTER TABLE is implicitly changing the value of 'AUTO_INCREMENT' from 'X' to 'Y'`.

vuvova commented on sql/sql_table.cc:
```diff
@@ -11370,6 +11413,7 @@ do_continue:;
       */
       table->file->ha_create_partitioning_metadata(alter_ctx.get_tmp_path(),
                                                    NULL, CHF_DELETE_FLAG);
+
```
remnant of previous fix versions? please delete

FarihaIS replied:
Thank you, deleted.

vuvova commented on sql/sql_table.cc:
```diff
@@ -127,6 +127,42 @@ static bool
+{
+  if (!(create_info->used_fields & HA_CREATE_USED_AUTO) ||
+      create_info->auto_increment_value == 0 ||
+      !table->found_next_number_field)
```
can `table->found_next_number_field` be 0 at this point?

FarihaIS replied:
I believe so -  for `ALTER TABLE t1 AUTO_INCREMENT=N `on a table without an `AUTO_INCREMENT` column, it can be null. Since InnoDB only populates `stats.auto_increment_value` when `found_next_number_field` is non-NULL, we'd risk comparing against stale stats without the check.

gkodinov commented:
There was no reply to Serg's review comments for 3 weeks now. Converting to draft. Please re-open when/if you resume working on this.

FarihaIS commented:
@gkodinov Apologies, not sure how I missed Serg's comments above, I'll work on this today and resubmit this once it's ready - thank you

FarihaIS commented on sql/sql_table.cc:
```diff
@@ -127,6 +127,42 @@ static bool
+static void
+check_auto_increment_lower_than_max(THD *thd,
+                                    const HA_CREATE_INFO *create_info,
+                                    TABLE *table)
```
Thanks, renamed to `check_auto_increment_lower_than_next`.

FarihaIS commented:
@gkodinov @vuvova thank you for waiting, I have addressed all comments now - could you please take a look and let me know if any other changes are needed?

gkodinov commented:
FYI: According to our development cycle we work on bugs In the following periods 15.03 - 30.04, 15.06 - 30.07, 15.12 - 31.01 and 15.09 - 30.10. So, please, expect to get a review somewhere between these two dates and the goal is to have your PR merged before the second date.
